### ISSUE TITLE
Reader for OSM data

### ISSUE THREAD

**Issue by simonellistonball** [harsha2010/magellan#29]
Open Street Map provides extensive and detailed public domain mapping data. It would be useful to be able to work with data from this source in Magellan. 

The primary format for exports of OSM data is an xml based format:
http://wiki.openstreetmap.org/wiki/OSM_XML

The planet.osm export in XML format is about 45G under bzip2 compression, making it hard to parse with a single file method, and hard to split, since XML tags cross tag boundaries in the canonical format of many of the tools, preventing split by lines. 

There is also a binary format which is based on Protocol Buffers:
http://wiki.openstreetmap.org/wiki/PBF_Format

This weighs in around 29G. It may be possible to extend existing protocol buffer input formats to work with PBF format, and use the dataframes API to join the nodes, ways and relations into a more magellan friendly format.

Both formats work on a schema based on a set of nodes (points with meta data), ways (sequences of nodes) and relations (collections of the above with additional metadata, also used to produce things like boundary polygons).

Many people use selected extracts in OSM or PBF format, making this a useful reader to include in magellan.

**Comment by harsha2010** [comment]
+1
Thanks Simon!

**Comment by adrianb3-telenav** [comment]
Hi guys,

I was wondering if you find useful also a PBF reader, in my experience, the OSM data consumers prefer this PBF format when processing continents or the whole planet.

I'm not a protocol buffer specialist but I know that the twitter guys were working with some protocol buffers input formats. 
https://github.com/twitter/elephant-bird/blob/master/examples/src/main/java/com/twitter/elephantbird/examples/ProtobufMRExample.java

Any idea if a similar approach will work with OSM PBF's in Spark, and then wrapping it in magellan?

Thanks,
Adrian

**Comment by harsha2010** [comment]
@adrianb3-telenav , thanks this is very useful feedback. Per our offline discussion we'll start looking into supporting PBF. As you pointed out, splittability might be tricky, I'll dig into this a bit and update the thread with a viable approach once i have researched this.

**Comment by adrianulbona** [comment]
great, thanks! I will let you know if I also find something on this thread.

**Comment by adrianulbona** [comment]
this should be at least evaluated. 
https://github.com/plasmap/geow

**Comment by raazesh-sainudiin** [comment]
Yes, we have been playing with magellan in databricks (just Ram's EU Spark Summit Talk 'ditabricksified' here:  http://www.math.canterbury.ac.nz/~r.sainudiin/courses/ScalableDataScience/2016/magellanGeoSpatialDatabricks.html ) and OSM with its various tools via piped-RDDs and could really use an integrated OSM reader with magellan!

**Comment by harsha2010** [comment]
Resolved by #34

**Comment by lamastex** [comment]
Would love to try it out when I get a chance. One could mix OSM with live hate-speech on twitter for 'ideological street safety' one day from the maths and codes here: https://lamastex.github.io/scalable-data-science/sds/research/mep/ . Thanks Ram!
